addons/minio/template/base/install.sh

```diff
@@ -247,7 +271,344 @@ function allow_pvc_resize() {
+function minio_swap_fs_migration_hostpaths() {
+    local suffix
+    local bkp_location
+    suffix=$(< /dev/urandom tr -dc A-Za-z0-9 | head -c8)
+    bkp_location=$(printf "%s-%s" "$MINIO_HOSTPATH" "$suffix")
+    mv "$MINIO_HOSTPATH" "$bkp_location"
+    printf "${YELLOW}A backup of the original Minio data has been stored at %s${NC}\n" "$bkp_location"
+
+    mv "$MINIO_MIGRATION_HOSTPATH" "$MINIO_HOSTPATH"
+    cp -Rfp "$bkp_location/.kurl" "$MINIO_HOSTPATH/"
+}
+
+# minio_uses_fs_format verifies if minio uses the legacy fs format. greps the file /data/.minio.sys/format.json
+# from within the pod.
+function minio_uses_fs_format() {
+    local format_string
+    format_string=$(kubectl exec -n $MINIO_NAMESPACE deploy/minio -- cat /data/.minio.sys/format.json 2>/dev/null)
+    if [ -z "$format_string" ]; then
+        bail "Failed to read /data/.miniosys/format.json inside minio pod"
+    fi
+    echo "$format_string" | grep -q '"format":"fs"'
+}
+
+# minio_svc_has_no_endpoints validates that the provided service in the provided namespace has no endpoints.
+function minio_svc_has_no_endpoints() {
+    local namespace=$1
+    local service=$2
+    kubectl get endpoints --no-headers -n "$namespace" "$service" -o custom-columns=:subsets | grep -q none
+}
+
+# minio_disable_minio_svc disables the minio service by tweaking its service selector.
+function minio_disable_minio_svc() {
+    kubectl patch svc -n "$MINIO_NAMESPACE" minio -p '{"spec":{"selector":{"app":"does-not-exist"}}}'
+}
+
+# minio_enable_minio_svc sets the minio service selector back to its default.
+function minio_enable_minio_svc() {
+    kubectl patch svc -n "$MINIO_NAMESPACE" minio -p '{"spec":{"selector":{"app":"minio"}}}'
+}
+
+# minio_prepare_volumes_for_migration sets the retention policy of the volumes to "retain" for both minio and its migration deployment.
+function minio_prepare_volumes_for_migration() {
+    local minio_pv=$1
+    local migration_pv=$2
+    # set both of the pv retention policies to 'retain'.
+    kubectl patch pv -n "$MINIO_NAMESPACE" "$minio_pv" -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'
+    kubectl patch pv -n "$MINIO_NAMESPACE" "$migration_pv" -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'
+}
+
+# minio_has_enough_space_for_fs_migration verifies if there is enough disk space in the cluster to execute the migration.
+function minio_has_enough_space_for_fs_migration() {
+    if [ -n "$MINIO_HOSTPATH" ]; then
+        minio_ask_user_hostpath_for_migration
+        return 0
+    fi
+    "$DIR"/bin/kurl cluster check-free-disk-space --debug --openebs-image "$KURL_UTIL_IMAGE" --bigger-than "$MINIO_CLAIM_SIZE" 2>&1
+}
+
+
+# minio_pods_are_finished returns 0 when there is no more pods with label app=minio in the minio namespace.
+function minio_pods_are_finished() {
+    local pods_count
+    pods_count=$(kubectl get pods -n "$MINIO_NAMESPACE" -l app=minio 2>/dev/null |wc -l)
```

Comment:
perhaps something like this to not falsely return 0 if the command failed:

```suggestion
    local pods
    pods="$(kubectl -n "$MINIO_NAMESPACE" get pods --no-headers -l app=minio 2>/dev/null)"
    if [ "$?" != "0" ]; then
        return 1
    fi
    local pods_count
    pods_count="$(echo "$pods" | wc -l)"
```

Reply:
i count resources convenience function would be nice come to think of it so others dont make the same mistake

Reply:
I have adapted your changes. What do you meant by your second comment here ? I could not follow.